fd_linux.go

```diff
@@ -1,45 +1,213 @@
-// calls or swap them out for tests.
-type fd interface {
-	Accept4(flags int) (fd, unix.Sockaddr, error)
+// A listenFD is a type that wraps a file descriptor used to implement
+// net.Listener.
+type listenFD interface {
+	io.Closer
+	EarlyClose() error
+	Accept4(flags int) (connFD, unix.Sockaddr, error)
 	Bind(sa unix.Sockaddr) error
-	Close() error
-	Connect(sa unix.Sockaddr) error
-	Getsockname() (unix.Sockaddr, error)
 	Listen(n int) error
-	NewFile(name string) *os.File
-	SetNonblock(nonblocking bool) error
+	Getsockname() (unix.Sockaddr, error)
+	SetNonblocking(name string) error
 }
 
-var _ fd = &sysFD{}
+var _ listenFD = &sysListenFD{}
 
-// sysFD is the system call implementation of fd.
-type sysFD struct {
-	fd int
+// A sysListenFD is the system call implementation of listenFD.
+type sysListenFD struct {
+	// These fields should never be non-zero at the same time.
+	fd int      // Used in blocking mode.
+	f  *os.File // Used in non-blocking mode.
 }
 
-func (fd *sysFD) Accept4(flags int) (fd, unix.Sockaddr, error) {
-	// Returns a regular file descriptor, must be wrapped in another
-	// sysFD for it to work properly.
-	nfd, sa, err := unix.Accept4(fd.fd, flags)
+// newListenFD creates a sysListenFD in its default blocking mode.
+func newListenFD() (*sysListenFD, error) {
+	fd, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sysListenFD{
+		fd: fd,
+	}, nil
+}
+
+// Blocking mode methods.
+
+func (lfd *sysListenFD) Bind(sa unix.Sockaddr) error         { return unix.Bind(lfd.fd, sa) }
+func (lfd *sysListenFD) Getsockname() (unix.Sockaddr, error) { return unix.Getsockname(lfd.fd) }
+func (lfd *sysListenFD) Listen(n int) error                  { return unix.Listen(lfd.fd, n) }
+
+func (lfd *sysListenFD) SetNonblocking(name string) error {
+	// From now on, we must perform non-blocking I/O, so that our
+	// net.Listener.Accept method can be interrupted by closing the socket.
+	if err := unix.SetNonblock(lfd.fd, true); err != nil {
+		return err
+	}
+
+	// Transition from blocking mode to non-blocking mode.
+	lfd.f = os.NewFile(uintptr(lfd.fd), name)
+
+	return nil
+}
+
+// EarlyClose is a blocking version of Close, only used for cleanup before
+// entering non-blocking mode.
+func (lfd *sysListenFD) EarlyClose() error { return unix.Close(lfd.fd) }
+
+// Non-blocking mode methods.
+
+func (lfd *sysListenFD) Accept4(flags int) (connFD, unix.Sockaddr, error) {
+	rc, err := lfd.f.SyscallConn()
 	if err != nil {
 		return nil, nil, err
 	}
 
-	return &sysFD{fd: nfd}, sa, nil
+	var (
+		nfd int
+		sa  unix.Sockaddr
+	)
+
+	doErr := rc.Read(func(fd uintptr) bool {
+		nfd, sa, err = unix.Accept4(int(fd), flags)
+
+		// When the socket is in non-blocking mode, we might see
+		// EAGAIN and end up here. In that case, return false to
+		// let the poller wait for readiness. See the source code
+		// for internal/poll.FD.RawRead for more details.
+		return err != unix.EAGAIN
```

Comment:
Can you ever get ECONNABORTED here? `internal/poll` checks for it: https://golang.org/src/internal/poll/fd_unix.go#L366

Seems unlikely at first sight, given the nature of vsock, but perhaps worth thinking about.

Reply:
Fixed!